WiniUI3 Setup server communication

herve-er opened the pull request:
## ✨ Pull Request Overview

This PR introduces the **communication architecture** for the **WinUI3 client**, enabling structured and reliable interactions with the server.

It establishes a **socket-based communication protocol** and a **dedicated service layer**, ensuring maintainability, scalability, and thread-safe data handling across the application.

---

### 🚀 Key Changes

* **Communication protocol**

  * Implemented a socket-based client for live communication with the server
  * Added a mock implementation for offline testing and development

* **Server communication service layer**

  * Introduced a centralized service for message handling and data synchronization
  * Abstracted server protocol details from higher-level components

* **View model updates**

  * Refactored all view models to use **thread-safe property setters**
  * Improved UI responsiveness and state consistency


depends on #1105

herve-er committed:
feat(Settings page WinUI3): Setup layer

herve-er committed:
fix(userinfo): fix avatar image conversion to CommBLOB

Added QImageWriter and QBuffer includes for better image processing. Updated UserInfo::toDynamicStruct to save the avatar in PNG format using QBuffer, ensuring proper formatting before conversion to CommBLOB.

herve-er committed:
fix(userinfo): fix avatar image conversion to CommBLOB

Added QImageWriter and QBuffer includes for better image processing. Updated UserInfo::toDynamicStruct to save the avatar in PNG format using QBuffer, ensuring proper formatting before conversion to CommBLOB.

herve-er committed:
refactor: Enhance code structure and server communication

Refactor the Infomaniak kDrive application to improve code organization and maintainability. Key changes include:
- Introduced dependency injection with `IServiceCollection` and `IServiceProvider`.
- Added `IServerCommProtocol` and `IServerCommService` interfaces for better server communication abstraction.
- Implemented signal handling for dynamic user updates.
- Created mock data classes for testing without a live server.
- Ensured UI thread safety with `UISafeObservableObject`.
- Added `IsNullToBoolOrVisibilityConverter` for flexible data binding.
- Cleaned up code by removing unused namespaces and redundant code.
- Enhanced error handling and logging throughout the application.
- Updated XAML files to align with new data binding logic.
- Introduced new classes like `Account`, `ServerInterface`, and `CommShared`.
- Modified project configuration files to include new dependencies.

herve-er committed:
refactor: Enhance code structure and server communication

Refactor the Infomaniak kDrive application to improve code organization and maintainability. Key changes include:
- Introduced dependency injection with `IServiceCollection` and `IServiceProvider`.
- Added `IServerCommProtocol` and `IServerCommService` interfaces for better server communication abstraction.
- Implemented signal handling for dynamic user updates.
- Created mock data classes for testing without a live server.
- Ensured UI thread safety with `UISafeObservableObject`.
- Added `IsNullToBoolOrVisibilityConverter` for flexible data binding.
- Cleaned up code by removing unused namespaces and redundant code.
- Enhanced error handling and logging throughout the application.
- Updated XAML files to align with new data binding logic.
- Introduced new classes like `Account`, `ServerInterface`, and `CommShared`.
- Modified project configuration files to include new dependencies.

herve-er committed:
feat(Comm WinUI3): Simplify code and remove unused methods

- Removed `ServerInterface` methods, retaining mock data handling.
- Eliminated `_hasError` field and related logic in `Drive`, streamlining error management.
- Removed `Reload` method in `Sync.

herve-er committed:
feat(Comm WinUI3): Simplify code and remove unused methods

- Removed `ServerInterface` methods, retaining mock data handling.
- Eliminated `_hasError` field and related logic in `Drive`, streamlining error management.
- Removed `Reload` method in `Sync.

herve-er commented on src/server/comm/guijobs/unknownrequestjob.cpp:
```diff
@@ -0,0 +1,64 @@
+    _exitInfo = ExitCode::LogicError;
+    _error = "unknown_request";
+    _errorDescr = "The request is unknown by the server application.";
+    return _exitInfo;
```
`_exitInfo` can safely be converted to bool

herve-er replied:
Done

herve-er committed:
chore(WinUI3): fix typo

herve-er committed:
chore(WinUI3): fix typo

herve-er commented on src/gui4/windows/kDrive client/kDrive client/ServerCommunication/Services/MockServerCommProtocol.cs:
```diff
@@ -0,0 +1,287 @@
+            {
+                Type = CommMessageType.Request,
+                Id = (int)NextId,
+                RequestNum = RequestNum.AccountInfoList,
```
Done

herve-er commented on src/gui4/windows/kDrive client/kDrive client/ServerCommunication/Services/SocketServerCommProtocol.cs:
```diff
@@ -0,0 +1,308 @@
+                    }
+                    else
+                    {
+                        Logger.Log(Logger.Level.Debug, $"Found request(Id) {requestId}, but the respons is not available yet null");
```
Done

herve-er commented on src/gui4/windows/kDrive client/kDrive client/ServerCommunication/Services/SocketServerCommProtocol.cs:
```diff
@@ -0,0 +1,308 @@
+                    Logger.Log(Logger.Level.Debug, $"No request(Id) found for {requestId}");
+                }
+                await Task.Delay(10).ConfigureAwait(false);
+                if (canceletionToken.IsCancellationRequested)
```
Done

herve-er committed:
chore(WinUI3): fix typos and formatting

herve-er committed:
chore(WinUI3): fix typos and formatting

herve-er committed:
docs: Update README.md with kDrive overview and JSON model

Added an overview of the kDrive application, including its components and communication methods.
Introduced a JSON Message Model section and detailed JSON representations for UserInfo, AccountInfo,
DriveInfo, and SyncInfo. Included a Requests list for the `UserInfoList` request.

herve-er committed:
docs: Update README.md with kDrive overview and JSON model

Added an overview of the kDrive application, including its components and communication methods.
Introduced a JSON Message Model section and detailed JSON representations for UserInfo, AccountInfo,
DriveInfo, and SyncInfo. Included a Requests list for the `UserInfoList` request.

herve-er committed:
refactor: change userDbId type to DbId

Updated the userDbId variable type from int to DbId to better represent the database identifier.

herve-er committed:
refactor: change userDbId type to DbId

Updated the userDbId variable type from int to DbId to better represent the database identifier.

herve-er committed:
refactor(UnknownRequestJob): update method return types

Changed return types of methods in `UnknownRequestJob` from `bool` to `ExitInfo` for better error handling.
Added `hasError` parameter to `serializeOutputParms`. Updated error reporting in `process` method.

herve-er committed:
refactor(UnknownRequestJob): update method return types

Changed return types of methods in `UnknownRequestJob` from `bool` to `ExitInfo` for better error handling.
Added `hasError` parameter to `serializeOutputParms`. Updated error reporting in `process` method.

herve-er committed:
Merge branch 'WinUI3-Setup-server-communication' of https://github.com/Infomaniak/desktop-kDrive into WinUI3-Setup-server-communication

herve-er committed:
chore(WinUI3 Comm): Fix warnings

herve-er committed:
docs: Update API documentation and JSON models

- Added a warning for base64-encoding of strings.
- Modified `UserInfo` model to specify PNG binary for `avatar`.
- Updated `Drive` model to define color format.
- Introduced `SynchronizationConfiguration` section with a note on structure changes.
- Replaced `UserInfoList` request with `LoginRequestToken` for OAuth2 integration.
- Updated successful and failure responses for `LoginRequestToken`.
- Added new signals `UserAdded` and `UserUpdated` with JSON structures.

herve-er committed:
chore(config): update file paths for communication settings

Updated file paths for storing communication configuration files in C# and C++.
Changed from user profile directory to local application data directory, renaming files to `.comm`.

herve-er committed:
docs: Update README.md for improved structure and clarity

Enhance the README.md with a Table of Contents, reformat sections, and clarify descriptions. JSON examples are now more readable, and terminology has been standardized. The Requests and Signals sections have been reorganized for better accessibility.

herve-er committed:
docs: standardize section headings in README.md

Updated the summary headings for `LoginRequestToken`, `UserAdded`, and `UserUpdated` by removing emoji icons to enhance documentation consistency.

herve-er committed:
Merge branch 'develop' into WinUI3-Setup-server-communication

# Conflicts:
#	src/libcommon/info/userinfo.cpp
#	src/server/CMakeLists.txt
#	src/server/comm/guijobs/guijobfactory.cpp
#	test/server/CMakeLists.txt

herve-er committed:
docs(GuiComm): Add doc for UserDbIdList

herve-er committed:
fix: Update logging condition in registry function

Modified the logging condition in `Utility::registryWalkSubKeys` to log only on failled key opening.

herve-er committed:
chore: Fix merge.

ClementKunz commented on src/gui4/windows/kDrive client/kDrive client/Utility/UISafeObservableObject.cs:
```diff
@@ -0,0 +1,74 @@
+
+        }
+
+        protected void OnPropertyChangedInUIThread([CallerMemberName] string? propertyName = null)
```
Couldn't simply overload `OnPropertyChanged` in order to avoid changing all the call to this method everywhere?

herve-er replied:
`OnPropertyChanged` isn’t virtual and therefore cannot be properly overridden. At best, it could be hidden by another method with the same name, which would generate a warning. Even though this warning can be suppressed using the `new` keyword, I don’t think it’s good practice to hide such a common function.

I prefer to keep it explicit for the caller whether they’re invoking the standard `OnPropertyChanged` method or the version that runs on the UI thread. Some private properties are not related to the UI and therefore don’t need the overhead of `OnPropertyChangedInUIThread`; they can safely be changed from any thread.

ClementKunz commented on src/gui4/windows/kDrive client/kDrive client/Utility/UISafeObservableObject.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace Infomaniak.kDrive
+{
+    public class UISafeObservableObject : ObservableObject
```
Can you briefly explain why you needed to overload this class?
Why those properties are changed in UI thread (causing some concurrency issues if I understand correctly)?

herve-er replied:
Comment added in code base ->

```
/// <summary>
/// Provides a thread-safe version of <see cref="ObservableObject"/> that ensures
/// property change notifications are always raised on the UI thread when using
/// <see cref="SetPropertyInUIThread"/>.
///
/// This class is particularly useful for view models whose properties are bound
/// to UI elements but may be updated from background threads. Since UI elements
/// can only be modified from the UI thread, this class helps prevent cross-thread
/// access issues and simplifies synchronization logic.
///
/// By using <see cref="SetPropertyInUIThread"/>, you avoid the need to manually
/// dispatch property change notifications to the UI thread in every background
/// update scenario.
///
/// For properties that are not related to the UI, the regular
/// <see cref="ObservableObject.SetProperty"/> method can still be used to avoid
/// unnecessary thread marshalling overhead.
/// </summary>
```

ClementKunz commented on src/gui4/windows/kDrive client/kDrive client/ViewModels/App/User.cs:
```diff
@@ -17,132 +17,141 @@
+                {
+                    AppModel.UIThreadDispatcher.TryEnqueue(async () =>
+                    {
+                        _avatarImageSource = await ByteArrayToImageSource(_avatar); // ByteArrayToImageSource need to be run in the UI thread
```
Why? This sounds like a potentially heavy task that could lock temporary the IU thread?

herve-er replied:
ImageSource can only be constructed on the UI thread.

However, the lambda is async and `ByteArrayToImageSource` is awaited, so it will run on a different thread, keeping the UI thread free to do other work and preventing it from freezing.

herve-er committed:
docs: Add documentation to UISafeObservableObject

Enhanced the documentation for UISafeObservableObject to explain its thread-safe functionality and usage.

herve-er committed:
refactor(user): simplify image data writing to stream

Updated the image data writing process in `User.cs` to use
`WriteAsync` for improved readability and efficiency.
Removed the use of `DataWriter` and streamlined the code.

herve-er committed:
Merge branch 'develop' into WinUI3-Setup-server-communication
